File: rofunc/planning/src/robotics-codes-from-scratch-master/matlab/MP.m
%%    Movement primitives with various basis functions
%%
%%    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
%%    Written by Sylvain Calinon <https://calinon.ch>
%%
%%    This file is part of RCFS.
%%
%%    RCFS is free software: you can redistribute it and/or modify
%%    it under the terms of the GNU General Public License version 3 as
%%    published by the Free Software Foundation.
%%
%%    RCFS is distributed in the hope that it will be useful,
%%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
%%    GNU General Public License for more details.
%%
%%    You should have received a copy of the GNU General Public License
%%    along with RCFS. If not, see <http://www.gnu.org/licenses/>.

function MP

%% Parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
param.nbFct = 12; %Number of basis functions (odd number for Fourier basis functions)
param.nbVar = 2; %Dimension of position data (here: x1,x2)
param.nbData = 100; %Number of datapoints in a trajectory
param.basisName = 'RBF'; %PIECEWISE, RBF, BERNSTEIN, FOURIER


%% Load handwriting data
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
demos=[];
load('../data/2Dletters/S.mat'); %Planar trajectories for writing alphabet letters
x = spline(1:size(demos{1}.pos,2), demos{1}.pos, linspace(1,size(demos{1}.pos,2),param.nbData))(:); %Resampling

if isequal(param.basisName,'FOURIER')
	%Fourier basis functions require additional symmetrization (mirroring) if the signal is a discrete motion 
	X = reshape(x, param.nbVar, param.nbData);
	X = [X, fliplr(X)]; %Build real and even signal, characterized by f(-x) = f(x)
	x = X(:); 
	param.nbData = param.nbData * 2;
end


%% Movement primitives encoding and decoding 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if isequal(param.basisName,'PIECEWISE')
    phi = buildPhiPiecewise(param.nbData, param.nbFct);
elseif isequal(param.basisName,'RBF')
    phi = buildPhiRBF(param.nbData, param.nbFct);
elseif isequal(param.basisName,'BERNSTEIN')
    phi = buildPhiBernstein(param.nbData, param.nbFct);
elseif isequal(param.basisName,'FOURIER')
    phi = buildPhiFourier(param.nbData, param.nbFct);
end

Psi = kron(phi, eye(param.nbVar)); %Transform to multidimensional basis functions
w = (Psi' * Psi + eye(param.nbVar*param.nbFct).*realmin) \ Psi' * x; %Estimation of superposition weights from data
x_hat = Psi * w; %Reconstruction of data

if isequal(param.basisName,'FOURIER')
	%Fourier basis functions require de-symmetrization of the signal after processing (for visualization)
	param.nbData = param.nbData/2;
	x = x(1:param.nbData*param.nbVar);
	x_hat = x_hat(1:param.nbData*param.nbVar);
	phi = real(phi); %for plotting
	%Psi = Psi(1:param.nbData*param.nbVar,:);
end


%% Plots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
h = figure('position',[10 10 1800 500]); 
clrmap = lines(param.nbFct);

%Plot signals
subplot(1,3,1); hold on; axis off; 
l(1) = plot(x(1:2:end,:), x(2:2:end,:), 'linewidth',2,'color',[.2 .2 .2]);
l(2) = plot(x_hat(1:2:end,:), x_hat(2:2:end,:), 'linewidth',2,'color',[.9 .0 .0]);
legend(l,{'Demonstration','Reproduction'},'fontsize',20); 
axis tight; axis equal; 

%Plot basis functions (display only the real part for Fourier basis functions)
subplot(1,3,2); hold on; 
for i=1:param.nbFct
	plot(phi(:,i), 'linewidth',2,'color',clrmap(i,:));
end
if isequal(param.basisName,'FOURIER')
	plot([param.nbData,param.nbData], [-1,1], 'k:','linewidth',4);
end
axis tight;
xlabel('t','fontsize',26); ylabel('\phi_k','fontsize',26);

%Plot Psi*Psi' matrix (covariance matrix at trajectory level)
subplot(1,3,3); hold on; axis off; title('\Psi\Psi^T','fontsize',26);
msh = [0 1 1 0 0; 0 0 1 1 0];
colormap(flipud(gray));
imagesc(abs(Psi * Psi'));
plot(msh(1,:)*size(Psi,1), msh(2,:)*size(Psi,1), 'k-');
if isequal(param.basisName,'FOURIER')
	plot(msh(1,:)*size(Psi,1)/2, msh(2,:)*size(Psi,1)/2, 'k:','linewidth',4);
end
axis tight; axis square; axis ij;

waitfor(h);
end

%Building piecewise constant basis functions
function phi = buildPhiPiecewise(nbData, nbFct) 
	phi = kron(eye(nbFct), ones(ceil(nbData/nbFct),1));
	phi = phi(1:nbData,:);
end

%Building radial basis functions (RBFs)
function phi = buildPhiRBF(nbData, nbFct) 
	t = linspace(0, 1, nbData);
	
%	tMu = linspace(t(1), t(end), nbFct); %Repartition of centers within input range
%	sigma = 1 / nbFct; %Standard deviation
	
	tMu = linspace(t(1)-1/(nbFct-3), t(end)+1/(nbFct-3), nbFct); %Repartition of centers to limit border effects
	sigma = 1 / (nbFct-2); %Standard deviation
	
	phi = exp(-(t' - tMu).^2 / sigma^2);
	%D = t' * ones(1,nbFct) - ones(nbData,1) * tMu;
	%phi = exp(-D.^2 / sigma^2);

	%Optional rescaling
	%phi = phi ./ repmat(sum(phi,2), 1, nbFct); 
end

%Building Bernstein basis functions
function phi = buildPhiBernstein(nbData, nbFct)
	t = linspace(0, 1, nbData);
	phi = zeros(nbData, nbFct);
	for i=1:nbFct
		phi(:,i) = factorial(nbFct-1) ./ (factorial(i-1) .* factorial(nbFct-i)) .* (1-t).^(nbFct-i) .* t.^(i-1);
	end
end

%Building Fourier basis functions
function phi = buildPhiFourier(nbData, nbFct)
	t = linspace(0, 1, nbData);

%	%Computation for general signals (incl. complex numbers)
%	d = ceil((nbFct-1)/2);
%	k = -d:d;
%	phi = exp(t' * k * 2 * pi * 1i); 
%	%phi = cos(t' * k * 2 * pi); %Alternative computation for real signal
		
	%Alternative computation for real and even signal
	k = 0:nbFct-1;
	phi = cos(t' * k * 2 * pi);
	%phi(:,2:end) = phi(:,2:end) * 2;
	%invPhi = cos(k' * t * 2 * pi) / nbData;
	%invPsi = kron(invPhi, eye(param.nbVar));	
end

File: rofunc/planning/src/robotics-codes-from-scratch-master/python/MP.py
'''
    Movement primitives applied to a 2D trajectory

    Copyright (c) 2021 Idiap Research Institute, http://www.idiap.ch/
    Written by Jeremy Maceiras <jeremy.maceiras@idiap.ch>,
    Sylvain Calinon <https://calinon.ch>

    This file is part of RCFS.

    RCFS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 3 as
    published by the Free Software Foundation.

    RCFS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with RCFS. If not, see <http://www.gnu.org/licenses/>.
'''

import numpy as np
import matplotlib.pyplot as plt
from math import factorial
import os
from pathlib import Path

# Building piecewise constant basis functions
def build_phi_piecewise(nb_data, nb_fct):
    phi = np.kron( np.identity(nb_fct) , np.ones((int(np.ceil(nb_data/nb_fct)),1)) )
    return phi[:nb_data]

# Building radial basis functions (RBFs)
def build_phi_rbf(nb_data, nb_fct):
    t = np.linspace(0,1,nb_data).reshape((-1,1))
    tMu = np.linspace( t[0] , t[-1] , nb_fct )
    phi = np.exp( -1e2 * (t.T - tMu)**2 )
    return phi.T

# Building Bernstein basis functions
def build_phi_bernstein(nb_data, nb_fct):
    t = np.linspace(0,1,nb_data)
    phi = np.zeros((nb_data,nb_fct))
    for i in range(nb_fct):
        phi[:,i] = factorial(nb_fct-1) / (factorial(i) * factorial(nb_fct-1-i)) * (1-t)**(nb_fct-1-i) * t**i
    return phi

# Building Fourier basis functions
def build_phi_fourier(nb_data, nb_fct):

    t = np.linspace(0,1,nb_data).reshape((-1,1))

    # Alternative computation for real and even signal
    k = np.arange(0,nb_fct).reshape((-1,1))
    phi = np.cos( t.T * k * 2 * np.pi )
    return phi.T

# General param parameters
# ===============================
param = {
    "nbFct" : 7, # Number of basis functions
    "nbVar" : 2, # Dimension of position data 
    "nbData" : 200, # Number of datapoints in a trajectory
    "basisName": "FOURIER" # PIECEWISE, RBF, BERNSTEIN, FOURIER
}

# Dictionary to call the good method in function of the good basis name.
basis_functions_method = {
    "BERNSTEIN": build_phi_bernstein,
    "PIECEWISE": build_phi_piecewise,
    "RBF": build_phi_rbf,
    "FOURIER": build_phi_fourier
}

# Load handwriting data
FILE_PATH = os.path.dirname(os.path.realpath(__file__))
DATASET_LETTER_FILE = "../data/2Dletters/C.npy"
x = np.load(str(Path(FILE_PATH,DATASET_LETTER_FILE)))[0,:param["nbData"],:2]

t = np.linspace(0,1,param["nbData"]) # Time range

# Fourier basis functions require additional symmetrization (mirroring) if the signal is a discrete motion 
if param["basisName"]=="FOURIER":
    x = np.vstack(( x , x[::-1] ))
    param["nbData"] *= 2

x = x.flatten()

# Generate MP with various basis functions
# ============================================
phi = basis_functions_method[param["basisName"]](param["nbData"],param["nbFct"])


psi = np.kron(phi, np.identity(param["nbVar"])) # Transform to multidimensional basis functions
w = np.linalg.solve(psi.T @ psi + np.identity(param["nbVar"] * param["nbFct"]) * 1e-8, psi.T @ x ) # Estimation of superposition weights from data
x_hat = psi @ w # Reconstruction data

# Fourier basis functions require de-symmetrization of the signal after processing (for visualization)
if param["basisName"]=="FOURIER":
    param["nbData"] /= 2
    x = x[:int(param["nbData"]*param["nbVar"])]
    x_hat = x_hat[:int(param["nbData"]*param["nbVar"])] 


# Plotting
# =========
fig,axs = plt.subplots(3,1)
axs[0].plot(x[::2], x[1::2], c='black', label='Original' )
axs[0].plot(x_hat[::2], x_hat[1::2], c='r', label='Reproduced' )
axs[0].axis("off")
axs[0].axis("equal")
axs[0].legend()

axs[1].set_xticks([0,param["nbData"]])
axs[1].set_xticklabels(["0","T"])
for j in range(param["nbFct"]):
    axs[1].plot(phi[:,j])
axs[1].set_ylabel("$\phi_k$")

axs[2].matshow( np.real(psi @ psi.T) )
axs[2].axis("off")
axs[2].set_xticks([])
axs[2].set_yticks([])
axs[2].set_title(" $\Psi \Psi^T$ ",pad=5)

plt.show()
